### ISSUE TITLE
REST client doesn't send body in redirect request

### ISSUE THREAD

**Issue by jdussouillez** [quarkusio/quarkus#41751]
### Describe the bug

REST client doesn't send the body in the redirect request after a 307 redirect from a POST request.
It happens both with `RedirectHandler` and the new `AdvancedRedirectHandler`.

The RFC says nothing about the body, however it seems the body should be sent again in the redirect request.

https://developer.mozilla.org/fr/docs/Web/HTTP/Status/307
> The method and the body of the original request are reused to perform the redirected request.
> [...]
> The only difference between 307 and [302](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/302) is that 307 guarantees that the method and the body will not be changed when the redirected request is made.

https://http.dev/307
> The 307 Temporary Redirect status code works the same way that the [302 Found](https://http.dev/302) status code does, except that 307 Temporary Redirect status code specifically guarantees that the HTTP method and message body will not be changed in the follow-up HTTP request.

### Expected behavior

REST client send the original request body in the redirect request after a 307 response.

### Actual behavior

REST client doesn't the original request body in the redirect request after a 307 response.

### How to Reproduce?

Reproducer: https://github.com/jdussouillez/quarkus-rest-client-post-redirect-post/tree/quarkus-v3.13

- Run the server
- Run the client, it fails with HTTP 400 because the body is `null`

Using `curl` it works fine as the body is transferred to the redirect request.

### Quarkus version or git rev

3.12.1 + current main branch (7e98e77)

### Additional information

This issue was first reported on another issue: https://github.com/quarkusio/quarkus/issues/35126#issuecomment-2213814084

**Comment by geoand** [comment]
Interesting. This will take some investigating

**Comment by geoand** [comment]
@vietj when using `io.vertx.core.http.HttpClientBuilder#withRedirectHandler` isn't the original body supposed to be included?

**Comment by geoand** [comment]
@vietj ^

**Comment by jdussouillez** [comment]
Btw, I created branch [`quarkus-v3.19`](https://github.com/jdussouillez/quarkus-rest-client-post-redirect-post/tree/quarkus-v3.19) to have a more recent version of Quarkus.

**Comment by geoand** [comment]
@vietj any update on this?

**Comment by jdussouillez** [comment]
Behavior is still the same in Quarkus [v3.28.2](https://github.com/jdussouillez/quarkus-rest-client-post-redirect-post/tree/quarkus-v3.28)

**Comment by geoand** [comment]
@vietj ^

**Comment by jdussouillez** [comment]
Bumping this again. *[Reproducer with Quarkus v3.35](https://github.com/jdussouillez/quarkus-rest-client-post-redirect-post/tree/quarkus-v3.35)*

Sorry to insist, but the current implementation doesn't respect the convention for HTTP 307.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Status/307

> The method and the body of the original request are reused to perform the redirected request.
> [...]
> The difference between 307 and [302](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Status/302) is that 307 guarantees that the client **will not change** the request method and body when the redirected request is made.

**Comment by jponge** [comment]
Ok, we'd need a pure Vert.x reproducer to get started

**Comment by jponge** [comment]
I'll have a look on Friday

**Comment by jponge** [comment]
I can confirm that the `vertx-core` HTTP client only handles redirects with `GET` and `HEAD`.

While you can define a custom redirect handler, there is no way (or I am missing something very obvious) to plug a payload due to the redirect handler being a `Function<HttpClientResponse, Future<RequestOptions>>` (if you call `send(...)` then you get a `Future<HttpClientResponse>`).

Here is a minimal Vert.x 5 reproducer:

```java
package yolo;

import io.vertx.core.Future;
import io.vertx.core.VerticleBase;
import io.vertx.core.Vertx;
import io.vertx.core.http.*;
import io.vertx.core.json.Json;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.web.Router;
import io.vertx.ext.web.RoutingContext;
import io.vertx.ext.web.handler.BodyHandler;

public class Main extends VerticleBase {

    private HttpClientAgent httpClient;

    @Override
    public Future<?> start() {
        Router router = Router.router(vertx);

        router.route().handler(BodyHandler.create());
        router.get("/hello").handler(this::hello);
        router.post("/echo").handler(this::echo);
        router.post("/_echo").handler(this::redirectEcho);
        router.get("/self-echo").handler(this::selfEcho);
        router.get("/redirected-self-echo").handler(this::redirectedSelfEcho);

        this.httpClient = vertx.httpClientBuilder()
                .build();

        return vertx.createHttpServer()
                .requestHandler(router)
                .listen(8080);
    }

    private void redirectedSelfEcho(RoutingContext rc) {
        JsonObject payload = new JsonObject()
                .put("name", "Mr Bean")
                .put("drives", "Mini");

        httpClient.request(HttpMethod.POST, 8080, "localhost", "/_echo")
                .compose(request -> {

                    // Only handles GET / HEAD and we can't have a custom handler that passes a body again
                    // request.setFollowRedirects(true);
                    // request.redirectHandler(response -> {
                    //     String target = response.getHeader("Location");
                    //     return httpClient.request(HttpMethod.POST, 8080, "localhost", target);
                    // });

                    return request
                            .putHeader("Content-Type", "application/json")
                            .send(payload.encode());
                })
                .compose(response -> {
                    // Custom-built "redirect" handler, but it needs the payload to be available somehow
                    if (response.statusCode() == 307) {
                        String target = response.getHeader("Location");
                        return httpClient
                                .request(HttpMethod.POST, 8080, "localhost", target)
                                .compose(request -> request
                                        .putHeader("Content-Type", "application/json")
                                        .send(payload.encode()));
                    } else {
                        return Future.succeededFuture(response);
                    }
                })
                .compose(HttpClientResponse::body)
                .compose(buffer -> rc.response()
                        .putHeader("Content-Type", "application/json")
                        .end(buffer));
    }

    private void redirectEcho(RoutingContext rc) {
        rc.response().setStatusCode(307);
        rc.redirect("/echo");
    }

    private void selfEcho(RoutingContext rc) {
        httpClient.request(HttpMethod.POST, 8080, "localhost", "/echo")
                .compose(request -> {
                    JsonObject payload = new JsonObject()
                            .put("name", "Mr Bean")
                            .put("drives", "Mini");
                    return request
                            .putHeader("Content-Type", "application/json")
                            .send(payload.encode());
                })
                .compose(HttpClientResponse::body)
                .compose(buffer -> rc.response()
                        .putHeader("Content-Type", "application/json")
                        .end(buffer));
    }

    private void echo(RoutingContext rc) {
        if (!"application/json".equals(rc.request().getHeader("Content-Type"))) {
            rc.response()
                    .setStatusCode(400)
                    .end("Not a JSON payload");
        } else {
            JsonObject payload = rc.body().asJsonObject();
            rc.response()
                    .putHeader("Content-Type", "application/json")
                    .end(payload.encodePrettily());
        }
    }

    private void hello(RoutingContext rc) {
        rc.response().end("Hello!");
    }

    static void main() {
        Vertx vertx = Vertx.vertx();
        vertx.deployVerticle(new Main());
    }
}

```

@vietj @tsegismont am I missing anything? 🙏

**Comment by jponge** [comment]
Also see https://github.com/eclipse-vertx/vert.x/blob/master/vertx-core/src/main/java/io/vertx/core/http/impl/DefaultRedirectHandler.java#L38 for the default handler, it only handles `HEAD`/`GET`

**Comment by cescoffier** [comment]
So, I think it's done like this on purpose for security reasons. You don't want your content (potentially sensitive) being redirected anywhere. I guess we may need an explicit option to allow this (and likely with a target whitelist).

**Comment by rsvoboda** [comment]
https://quarkus.io/guides/rest-client#redirection should be also enhanced to explain the behavior.

**Comment by vietj** [comment]
it used to be possible to redirect with a body in a redirect handler, however we felt it was wiser to avoid this because of what @cescoffier said and also the fact that some streams can have an unknown size and this kind of solution requires buffering of the payload.

**Comment by vietj** [comment]
that being said we are open to allow an extended handler in vertx 5 to make this scenario possible if there is demand for it.

**Comment by cescoffier** [comment]
@geoand @jponge I think the rest client should allow (explicitly) to transfer the body. WDYT?

**Comment by geoand** [comment]
Yeah, it would be nice if Vert.x gave us that possibility so we could then integrate it into the REST Client

**Comment by jnbdz** [comment]
Following up on the Vert.x side first, as discussed above: eclipse-vertx/vert.x#6348 generalises the body buffering that Vert.x master already does for `QUERY` requests to every method that can carry a body. When the redirected request keeps the method, which is what a `RedirectHandler` or `AdvancedRedirectHandler` does for a `307`/`308`, the buffered body is sent again. The buffer is bounded by the existing `ClientRedirectConfig#setMaxBufferedSize` (4 KB by default); above that, redirections other than `303` are returned to the caller instead of being followed with an empty body, and a `303` is still followed as a `GET`.

The Vert.x default policy is not touched, so the explicit opt-in cescoffier asked for stays where it is today: only a client with a redirect handler that decides to keep the method sends the body again.

Once that lands in a Vert.x release Quarkus picks up, the REST Client needs no code change for the reporter's scenario. What I would add on the Quarkus side then is a `quarkus.rest-client.max-redirect-buffered-size` property mapped to that Vert.x setting, since uploads larger than 4 KB are the common case, plus a paragraph in the redirection section of the guide, as rsvoboda suggested.
